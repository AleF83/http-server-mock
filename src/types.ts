export interface ServerMockOptions {
  name: string;
  port: number;
}

export interface ServerMockInfo {
  id: string;
  name: string;
  port: number;
}

export class ServerResponse {
  url = '/';
}
