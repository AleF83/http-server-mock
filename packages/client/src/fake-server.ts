import { AxiosInstance } from 'axios';
import { RegisterResponseMockRequest, RegisterResponseMockResponse, FakeServerInfo } from 'http-server-mock-common';

export class FakeServerClient {
  constructor(private readonly _id: string, private readonly _axios: AxiosInstance) {}

  get id(): string {
    return this._id;
  }

  async start(): Promise<void> {
    const response = await this._axios.post<void>(`/servers/${this._id}/start`);
    return response.data;
  }

  async stop(): Promise<void> {
    const response = await this._axios.post<void>(`/servers/${this._id}/stop`);
    return response.data;
  }

  async getInfo(): Promise<FakeServerInfo> {
    const response = await this._axios.get<FakeServerInfo>(`/servers/${this._id}`);
    return response.data;
  }

  async registerResponseMock(responseMock: RegisterResponseMockRequest): Promise<RegisterResponseMockResponse> {
    const response = await this._axios.post<RegisterResponseMockResponse>(
      `/servers/${this._id}/register-response-mock`,
      responseMock
    );
    return response.data;
  }
}
