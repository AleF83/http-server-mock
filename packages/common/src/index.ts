export * from './request-matching';
export * from './fake-server';
export * from './server-calls';
