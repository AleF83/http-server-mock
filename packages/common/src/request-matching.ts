export type StringMatcher = string | string[] | RegExp;

export interface KeyValueMatcher {
  name: StringMatcher;
  value: StringMatcher;
}

export interface UrlMatcher {
  protocol?: StringMatcher;
  hostname?: StringMatcher;
  query?: StringMatcher | KeyValueMatcher[];
}

// TODO: Complete
export interface BodyMatcher<T = any> {
  mode: 'exact' | 'partial';
  partial?: Partial<T>;
}

export interface RequestMatcher<T = any> {
  url?: string | RegExp | UrlMatcher;
  headers?: { [k: string]: string } | KeyValueMatcher[];
  body?: T | BodyMatcher<T>;
}
