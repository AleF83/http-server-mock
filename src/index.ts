import startServer from './server';

const server = startServer();

server.listen(8080, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }
  console.log(`Server listening at ${address}`);
});
