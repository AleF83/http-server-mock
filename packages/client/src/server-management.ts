import { AxiosInstance } from 'axios';
import { CreateFakeServerRequest, CreateFakeServerResponse, FakeServerInfo } from 'http-server-mock-common';
import { FakeServerClient } from './fake-server';

export class ServerManagementClient {
  constructor(private readonly _axios: AxiosInstance) {}

  async createFakeServer(request: CreateFakeServerRequest): Promise<FakeServerClient> {
    const response = await this._axios.post<CreateFakeServerResponse>('', request);
    const fakeServerId = response.data.id;
    return new FakeServerClient(fakeServerId, this._axios);
  }

  async deleteFakeServer(id: string): Promise<void> {
    await this._axios.delete<void>(`/${id}`);
  }

  async listFakeServers(): Promise<FakeServerInfo[]> {
    const response = await this._axios.get<FakeServerInfo[]>('');
    return response.data;
  }

  async getFakeServerClient(id: string): Promise<FakeServerClient> {
    const response = await this._axios.get<FakeServerInfo>(`/${id}`);
    return new FakeServerClient(response.data.id, this._axios);
  }
}
