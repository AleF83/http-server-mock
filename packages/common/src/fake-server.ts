import { RequestMatcher } from './request-matching';

export enum FakeServerStatus {
  Created = 'created',
  Initialized = 'initialized',
  Starting = 'starting',
  Running = 'running',
  Stopping = 'stopping',
  Stopped = 'stopped',
}

export interface FakeServerInfo {
  id: string;
  name: string;
  port: number;
  status: FakeServerStatus;
}

export interface CreateFakeServerRequest {
  name: string;
  port: number;
  startOnInit?: boolean;
}

export interface CreateFakeServerResponse {
  id: string;
  status: FakeServerStatus;
}

export interface ResponseExpectation<T = any> {
  headers?: { [key: string]: string };
  body?: T;
}

export interface ResponseBehavior {
  delay: number;
}

export interface RegisterResponseMockRequest {
  requestMatcher?: RequestMatcher;
  expectedResponse?: ResponseExpectation;
  behavior?: ResponseBehavior;
}

export interface RegisterResponseMockResponse {
  id: string;
}
