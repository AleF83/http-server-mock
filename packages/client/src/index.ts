import axios from 'axios';
import { ServerManagementClient } from './server-management';
export * from './server-management';
export * from './fake-server';

const DEFAULT_SERVER_MANAGER_PORT = '8080';

export function createServerManagementClient(baseUrl: string): ServerManagementClient {
  const url = new URL(baseUrl);
  if (!url.port) {
    url.port = DEFAULT_SERVER_MANAGER_PORT;
  }
  const axiosInstance = axios.create({ baseURL: `${url.origin}/servers` });
  return new ServerManagementClient(axiosInstance);
}
