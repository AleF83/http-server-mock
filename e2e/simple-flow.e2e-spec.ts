import axios, { AxiosInstance } from 'axios';
import { ServiceMockOptions, ServiceMockInfo } from '../src/types';

describe('Simple flow', () => {
  let id: string;
  const name = 'e2e service mock';
  const port = 10000;

  let client: AxiosInstance;

  beforeAll(() => {
    client = axios.create({ baseURL: 'http://localhost:8080' });
  });

  test('No mocks on init', async () => {
    const response = await client.get('/');
    expect(response.data).toHaveLength(0);
  });

  test('Create service mock', async () => {
    const serviceMockOptions: ServiceMockOptions = { name, port };
    const response = await client.post<{ id: string }>('/', serviceMockOptions);
    expect(response.status).toEqual(201);
    id = response.data.id;
    expect(id).toBeDefined();
  });

  test('Get service mock info', async () => {
    const response = await client.get<ServiceMockInfo>(`/${id}`);
    expect(response.status).toEqual(200);
    expect(response.data).toMatchSnapshot({
      id: expect.any(String),
    });
  });

  // test('Set response', () => {});

  // test('Call service', () => {});

  test('Delete service mock', async () => {
    const response = await client.delete<ServiceMockInfo>(`/${id}`);
    expect(response.status).toEqual(200);
    expect(response.data.id).toEqual(id);
    expect(response.data).toMatchSnapshot({
      id: expect.any(String),
    });
  });
});
