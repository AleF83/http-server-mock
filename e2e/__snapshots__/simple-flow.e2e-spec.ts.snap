// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple flow Delete fake server 1`] = `
Object {
  "id": Any<String>,
  "name": "e2e service mock",
  "port": 10000,
}
`;

exports[`Simple flow Get fake-server info 1`] = `
Object {
  "id": Any<String>,
  "name": "e2e service mock",
  "port": 10000,
}
`;
